sui-execution/latest/sui-adapter/src/static_programmable_transactions/typing/verify/hot_entry_arguments.rs

```diff
@@ -0,0 +1,488 @@
+    execution_status::CommandArgumentError,
+};
+
+/// Marks if a clique is hot or not
+#[derive(Clone, Copy, Debug)]
+enum Temperature {
+    /// The clique is always hot--since it's heat was not the result of a hot potato value
+    AlwaysHot,
+    /// The clique is hot because it contains hot potato values--tracked via the `usize` count
+    Count(usize),
+}
+
+/// The data for a clique, which tracks the what values/locations are used together. Once a value
+/// is used with another in a `Command`, their cliques are merged forever.
+enum Clique {
+    Merged(CliqueID),
+    Root(Temperature),
+}
+
+type CliqueID = usize;
+
+#[must_use]
+enum Value {
+    // We don't want to entangle the TxContext with cliques
+    TxContext,
+    Normal {
+        /// The clique this value belongs to
+        clique: CliqueID,
+        /// If the value contributes to the hot-value count in its clique
+        heats: bool,
+    },
+}
+
+/// A list of all cliques. The size is bound by the number of inputs + number of commands
+struct Cliques(Vec<Clique>);
+
+struct Context {
+    cliques: Cliques,
+    tx_context: Option<Value>,
+    gas_coin: Option<Value>,
+    objects: Vec<Option<Value>>,
+    pure: Vec<Option<Value>>,
+    receiving: Vec<Option<Value>>,
+    results: Vec<Vec<Option<Value>>>,
+}
+
+impl Temperature {
+    fn add(self, other: Temperature) -> Result<Temperature, ExecutionError> {
+        Ok(match (self, other) {
+            (Temperature::AlwaysHot, _) | (_, Temperature::AlwaysHot) => Temperature::AlwaysHot,
+            (Temperature::Count(a), Temperature::Count(b)) => Temperature::Count(
+                a.checked_add(b)
+                    .ok_or_else(|| make_invariant_violation!("Hot count overflow"))?,
+            ),
+        })
+    }
+
+    fn sub(self, b: usize) -> Result<Temperature, ExecutionError> {
+        Ok(match self {
+            Temperature::AlwaysHot => Temperature::AlwaysHot,
+            Temperature::Count(a) => Temperature::Count(
+                a.checked_sub(b)
+                    .ok_or_else(|| make_invariant_violation!("Hot count cannot go negative"))?,
+            ),
+        })
+    }
+
+    /// Returns true if the clique is always hot or has a positive hot count
+    fn is_hot(&self) -> bool {
+        match self {
+            Temperature::AlwaysHot => true,
+            Temperature::Count(c) => *c > 0,
+        }
+    }
+}
+
+impl Cliques {
+    /// Creates a new empty set of cliques
+    fn new() -> Self {
+        Self(vec![])
+    }
+
+    /// Creates a new clique and returns its ID
+    fn next(&mut self) -> CliqueID {
+        let id = self.0.len();
+        self.0.push(Clique::Root(Temperature::Count(0)));
+        id
+    }
+
+    /// Returns the root of the clique (resolving any merges/forwards)
+    fn root(&self, id: CliqueID) -> CliqueID {
+        match &self.0[id] {
+            Clique::Root(_) => id,
+            Clique::Merged(index) => self.root(*index),
```

Comment:
How big can the number of cliques grow? I think they can grow to be larger than enough to possibly cause an issue due to this recursive call. So I think it would probably be best to write this imperatively (🙁)